torch/distributed/distributed_c10d.py

```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            process will block and wait for collectives to complete before
+            throwing an exception. When ``NCCL_ASYNC_ERROR_HANDLING`` is set,
+            this is the duration after which collectives will be aborted
+            asynchronously and the process will crash. ``NCCL_BLOCKING_WAIT``
+            will provide errors to the user which can be caught and handled,
+            but due to its blocking nature, it has a performance overhead. On
+            the other hand, ``NCCL_ASYNC_ERROR_HANDLING`` has very little
+            performance overhead, but crashes the process on errors. This is
+            done since CUDA execution is async and it is no longer safe to
+            continue executing user code since failed async NCCL operations
+            might result in subsequent CUDA operations running on corrupted
+            data. Only one of these two environment variables should be set.
+        backend (str or Backend, optional): The backend to use. Depending on
+            build-time configurations, valid values are ``gloo`` and ``nccl``.
+            By default uses the same backend as the global group. This field
+            should be given as a lowercase string (e.g., ``"gloo"``), which can
+            also be accessed via :class:`Backend` attributes (e.g.,
+            ``Backend.GLOO``). If ``None`` is passed in, the backend
+            corresponding to the default process group will be used. Default is
+            ``None``.
+        pg_options (ProcessGroupOptions, optional): process group options
+            specifying what additional options need to be passed in during
+            the construction of specific process groups. i.e. for the ``nccl``
+            backend, is_high_priority_stream can be specified so that process
+            group can pick up high priority cuda streams.
+
+    Returns:
+        The subgroup containing the current rank, and all the subgroups used for cleanup.
+    """
+    subgroups = []
+    cur_subgroup = None
+    if ranks_per_subgroup_list:
+        for ranks in ranks_per_subgroup_list:
+            subgroup = new_group(
+                ranks=ranks, timeout=timeout, backend=backend, pg_options=pg_options
+            )
+            subgroups.append(subgroup)
+
+            rank = get_rank()
+            if rank in ranks:
+                cur_subgroup = subgroup
+                logger.info("Rank {} is assigned to subgroup {}".format(rank, ranks))
+        return cur_subgroup, subgroups
+
+    # By default, create one subgroup per machine.
+    # Assume that each machine has the same number of devices.
+    num_devices_per_machine = (
+        torch.cuda.device_count()
+        if torch.cuda.is_available()
+        else (os.cpu_count() or 1)
+    )
+    # If the given main group does not use all the devices,
+    # then the number of devices used by the subgroup should exceed the world size.
+    num_devices_per_machine = min(get_world_size(group), num_devices_per_machine)
```

Comment:
Is this logic correct? Say there are 8 GPUs per machine and 4 machines. If `group` uses 4 GPUs across each machine, then `num_devices_per_machine=8`, but should it be 4 instead? Or are we indeed okay with still using all GPUs on each machine even though `group` doesn't (in which case we shouldn't need this check)?

Reply:
I decide to break this util function into two functions, one for equi-size partition, one for manually specifying subgroups.